### PR TITLE
Replace `Vue.extend` with `defineComponent` in `LabeledInput.vue`

### PR DESCRIPTION
**Pull request by rak-phillip**
### Summary
contributes to #10104 

Rewrote `compact-input.ts` and `labeled-form-element.ts` mixins as composables to address TypeScript issues when using mixins with `defineComponent`. The new composables are written to adhere to the conventions and best practices [^1] as best as possible, taking into consideration what's available to us in Vue 2.7.14. 

Moving to composables positions us for a smoother Vue 3 migration.

[^1]: https://vuejs.org/guide/reusability/composables.html#conventions-and-best-practices

### Areas or cases that should be tested

Any forms that utilize `LabeledInput`

### CODE REVIEW

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ describe('component: LabeledInput', () => {
   it('should emit input only once', () => {
     const value = '2';
     const delay = 1;
-    const wrapper = mount(LabeledInput, {
+    const wrapper = mount(LabeledInput as any, {
```
Comment (rak-phillip) [inline pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:9]: I'm not a fan of using `any` here, but I'm failing to find a solution that will satisfy vue test utils expectations here.

Comment (codyrancher) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:8]: It doesn't appear to cause any issues for me if you remove the type cast

image

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue [hunk 2]
```diff
@@ -90,24 +101,85 @@ export default (
...
+    required: {
+      type:    Boolean,
+      default: false,
+    },
+
+    disabled: {
+      type:    Boolean,
+      default: false,
+    },
+  },
```
Comment (codyrancher) [inline pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:152]: I think you mentioned this during our conversation. I don't think we want to have to define these props here or anywhere the composables are used. At the very least I think we should define the props in the composable files and allow the consumer to spread them in the object here. If there's a better solution we should do that instead.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue [hunk 3]
```diff
@@ -144,7 +216,7 @@ export default (
       if (this.type !== 'cron' || !this.value) {
         return;
       }
-      if (!isValidCron(this.value)) {
+      if (typeof this.value === 'string' && !isValidCron(this.value)) {
```
Comment (codyrancher) [inline pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:219]: This seems like it should be pushed into isValidCron. The function should enforce it's own type expectations.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue [hunk 4]
```diff
@@ -173,15 +245,22 @@ export default (
...
      * The max length for the Labeled Input.
      */
-    _maxlength(): number | null {
+    _maxlength(): number | string {
       if (this.type === 'text' && this.maxlength) {
         return this.maxlength;
       }
 
-      return null;
+      return '';
```
Comment (codyrancher) [inline pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:253]: Why this change?

File: shell/composables/useCompactInput.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { computed } from 'vue';
+import type { ComputedRef } from 'vue';
```
Comment (codyrancher) [inline shell/composables/useCompactInput.ts:2]: Is there a reason to keep these separate?

Comment (rak-phillip) [reply shell/composables/useCompactInput.ts:2]: It's note required, I tend to prefer to try and keep it clear when an import is only used for type annotations. I'm not opposed to merging the two imports.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue (hunk 4 above)
Comment (rak-phillip) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:253]: `InputHTMLAttributes.maxlength` on input has a type of `Numberish | undefined` (`Numberish` is just `number | string`).

I found that an empty string indicates an unlimited `maxlength` in my testing and felt that it was appropriate, but I suppose that we could opt for `undefined` instead of `null` and get the same outcome.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue (hunk 3 above)
Comment (rak-phillip) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:219]: I would agree, but `isValidCron` is provided by the `cron-validator` package.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue (hunk 2 above)
Comment (rak-phillip) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:152]: Yeah, I think it should be possible to do this. I was on the fence about implementing this right now, as I think that the PR is getting quite large. 

I don't think that taking this on will add too much complexity on top of what's represented here, so I'll give it a try.

Comment (rak-phillip) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue]: This was quite straightforward to perform and we still get prop types exposed in our templates. Let me know what you think!

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts (hunk 1 above)
Comment (rak-phillip) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:8]: It looks like the errors I'm receiving are reported by tsc. Can you let me know what the output of `yarn build:lib` is for you without the type cast?

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue [hunk 6]
```diff
@@ -90,24 +96,38 @@ export default (
...
+      isCompact,
+    };
   },
 
   data() {
     return {
       updated:          false,
-      validationErrors: ''
+      validationErrors: '',
+      focused:          false,
```
Comment (rak-phillip) [inline pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:126]: I think that `focused` needs to be exposed through the composable.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts (hunk 1 above)
Comment (codyrancher) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:8]: Yeah, running yarn build:lib fails. I've spent a heck of alot of time trying to make this work in a better way. I managed to get a variation of https://github.com/vuejs/vue-test-utils/issues/2026#issuecomment-1460894573 to partially work.

Unfortunately I had to pull in the entire test-utils definition in to fix other issues. I think using `any` is fine until we migrate to vue3 and a new version of test-utils.

**Review by codyrancher — APPROVED** [review]

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts (hunk 1 above)
Comment (codyrancher) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:8]: After posting this I did have one other thought... Why is the test getting built when running `yarn build:lib`? If it doesn't need to be bundled into the library I found this diff works fine

image

**Review by mantis-toboggan-md — APPROVED** [review]

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts (hunk 1 above)
Comment (rak-phillip) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:8]: Wouldn't that exclude all of our tests from the compilation process, which begs the question, what's the benefit of writing our tests in TypeScript in the first place?

I might opt for `any` for now and follow up in a separate PR to address the vue-test-utils issues. I'm seeing multiple github issues that appear to be related to the one we're experiencing here, https://github.com/vuejs/vue-test-utils/issues/2026 for example

Comment (codyrancher) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.test.ts:8]: I think the tests are already being excluded from the compilation process since yarn test:ci functions as already noted. I personally don't see the point of typescript until composition or vue 3 is done.

I'm fine with you making the change later anyway. I just wanted to bring up what I found.
